# PyCity School Analysis
## Module 4

In [293]:
#importing module
import pandas as pd


In [294]:
#Files have been read and stored into Pandas DataFrame
school_data = pd.read_csv("Resources/schools_complete.csv")
student_data = pd.read_csv("Resources/students_complete.csv")

#Files merged ino single dataset
school_data_all = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_all.head()


Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

# District Summary

In [295]:
#Total number of unique schools in the district
school_count = len(school_data_all["school_name"].unique())
school_count 

15

In [296]:
#Total number of students in the district
student_count = len(school_data_all["Student ID"].unique())
student_count

39170

In [297]:
#Total budget of schools in the district
schools_reduced = school_data_all.drop_duplicates("school_name") 
total_budget = schools_reduced.sum()["budget"]
total_budget

24649428

In [298]:
#Average math score of students in district 
average_math_score = school_data_all["math_score"].mean()
average_math_score

78.98537145774827

In [299]:
#Average reading score of students in district
average_reading_score = school_data_all["reading_score"].mean()
average_reading_score

81.87784018381414

In [300]:
#Percentage of students passing math in the district
passing_math_count = school_data_all[(school_data_all["math_score"] >= 70)].count()["student_name"]
passing_math_percentage = passing_math_count / float(student_count) * 100
passing_math_percentage

74.9808526933878

In [301]:
#Percentage of students passing reading in the district
passing_reading_count = school_data_all[(school_data_all["reading_score"] >= 70)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(student_count) * 100
passing_reading_percentage

85.80546336482001

In [302]:
##Percentage of students passing math and reading in the district
passing_math_reading_count = school_data_all[
    (school_data_all["math_score"] >= 70) & (school_data_all["reading_score"] >= 70)
].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(student_count) * 100
overall_passing_rate

65.17232575950983

In [303]:
#Creating DF of the district
district_summary = pd.DataFrame({"Total Schools": [school_count] , "Total Students": [student_count], "Total Budget": [total_budget],
                                  "Average Math Score": [average_math_score], "Average Reading Score": [average_reading_score], 
                                   "% Passing Math": [passing_math_percentage], "% Passing Reading": [passing_reading_percentage], 
                                   "% Overall Passing Rate": [overall_passing_rate]})

#Formatting and Mapping district dataframe
district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)
district_summary["Average Math Score"] = district_summary["Average Math Score"].map("{:,.3f}".format)
district_summary["Average Reading Score"] = district_summary["Average Reading Score"].map("{:,.3f}".format)
district_summary["% Passing Math"] = district_summary["% Passing Math"].map("{:,.3f}".format)
district_summary["% Passing Reading"] = district_summary["% Passing Reading"].map("{:,.3f}".format)
district_summary["% Overall Passing Rate"] = district_summary["% Overall Passing Rate"].map("{:,.3f}".format)

#District Dataframe displayed
district_summary.head()

Total Schools Total Students    Total Budget Average Math Score  \
0             15         39,170  $24,649,428.00             78.985   

  Average Reading Score % Passing Math % Passing Reading  \
0                81.878         74.981            85.805   

  % Overall Passing Rate  
0                 65.172

# School Summary

In [304]:
#Setting up index
school_types = school_data.set_index(["school_name"])["type"]

In [305]:
#Students Per School // 15 added in the head to include all schools
per_school_counts = school_data_all["school_name"].value_counts()
per_school_counts.head(15)

Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: school_name, dtype: int64

In [306]:
#Total budget per school // 15 added in the head to include all schools
per_school_budget = school_data_all.groupby(["school_name"]).mean(numeric_only=True)["budget"]
per_school_budget.head(15)

school_name
Bailey High School       3124928.0
Cabrera High School      1081356.0
Figueroa High School     1884411.0
Ford High School         1763916.0
Griffin High School       917500.0
Hernandez High School    3022020.0
Holden High School        248087.0
Huang High School        1910635.0
Johnson High School      3094650.0
Pena High School          585858.0
Rodriguez High School    2547363.0
Shelton High School      1056600.0
Thomas High School       1043130.0
Wilson High School       1319574.0
Wright High School       1049400.0
Name: budget, dtype: float64

In [307]:
#Total budget per student // 15 added in the head to include all schools
school_budget_per_student = per_school_budget / per_school_counts
school_budget_per_student.head(15)

Bailey High School       628.0
Cabrera High School      582.0
Figueroa High School     639.0
Ford High School         644.0
Griffin High School      625.0
Hernandez High School    652.0
Holden High School       581.0
Huang High School        655.0
Johnson High School      650.0
Pena High School         609.0
Rodriguez High School    637.0
Shelton High School      600.0
Thomas High School       638.0
Wilson High School       578.0
Wright High School       583.0
dtype: float64

In [308]:
# Math average score per school
per_school_math = school_data_all.groupby(["school_name"])["math_score"].mean()
per_school_math

school_name
Bailey High School       77.048432
Cabrera High School      83.061895
Figueroa High School     76.711767
Ford High School         77.102592
Griffin High School      83.351499
Hernandez High School    77.289752
Holden High School       83.803279
Huang High School        76.629414
Johnson High School      77.072464
Pena High School         83.839917
Rodriguez High School    76.842711
Shelton High School      83.359455
Thomas High School       83.418349
Wilson High School       83.274201
Wright High School       83.682222
Name: math_score, dtype: float64

In [309]:
#Reading average score per school
per_school_reading = school_data_all.groupby(["school_name"])["reading_score"].mean()
per_school_reading

school_name
Bailey High School       81.033963
Cabrera High School      83.975780
Figueroa High School     81.158020
Ford High School         80.746258
Griffin High School      83.816757
Hernandez High School    80.934412
Holden High School       83.814988
Huang High School        81.182722
Johnson High School      80.966394
Pena High School         84.044699
Rodriguez High School    80.744686
Shelton High School      83.725724
Thomas High School       83.848930
Wilson High School       83.989488
Wright High School       83.955000
Name: reading_score, dtype: float64

In [310]:
#Percentage of students passing math per school
school_passing_math =  school_data_all.loc[(school_data_all["math_score"] >= 70)]
groupm_per_school = school_passing_math["school_name"].value_counts()
school_passing_math_perc = groupm_per_school / per_school_counts * 100
school_passing_math_perc

Bailey High School       66.680064
Cabrera High School      94.133477
Figueroa High School     65.988471
Ford High School         68.309602
Griffin High School      93.392371
Hernandez High School    66.752967
Holden High School       92.505855
Huang High School        65.683922
Johnson High School      66.057551
Pena High School         94.594595
Rodriguez High School    66.366592
Shelton High School      93.867121
Thomas High School       93.272171
Wilson High School       93.867718
Wright High School       93.333333
Name: school_name, dtype: float64

In [311]:
#Percentage of students passing reading per school
school_passing_reading =  school_data_all.loc[(school_data_all["reading_score"] >= 70)]
groupr_per_school = school_passing_reading["school_name"].value_counts()
school_passing_reading_perc = groupr_per_school / per_school_counts * 100
school_passing_reading_perc

Bailey High School       81.933280
Cabrera High School      97.039828
Figueroa High School     80.739234
Ford High School         79.299014
Griffin High School      97.138965
Hernandez High School    80.862999
Holden High School       96.252927
Huang High School        81.316421
Johnson High School      81.222432
Pena High School         95.945946
Rodriguez High School    80.220055
Shelton High School      95.854628
Thomas High School       97.308869
Wilson High School       96.539641
Wright High School       96.611111
Name: school_name, dtype: float64

In [312]:
#Percentage of students passing both reading and math per school
passing_math_and_reading = school_data_all[
    (school_data_all["reading_score"] >= 70) & (school_data_all["math_score"] >= 70)]
groupb_per_school = passing_math_and_reading["school_name"].value_counts()
school_passing_both_perc = groupb_per_school / per_school_counts * 100
school_passing_both_perc

Bailey High School       54.642283
Cabrera High School      91.334769
Figueroa High School     53.204476
Ford High School         54.289887
Griffin High School      90.599455
Hernandez High School    53.527508
Holden High School       89.227166
Huang High School        53.513884
Johnson High School      53.539172
Pena High School         90.540541
Rodriguez High School    52.988247
Shelton High School      89.892107
Thomas High School       90.948012
Wilson High School       90.582567
Wright High School       90.333333
Name: school_name, dtype: float64

In [313]:
#Creating Df of school summary 
per_school_summary = pd.DataFrame({"School Type": school_types, "Total Students": per_school_counts, 
                                   "Total School Budget": per_school_budget,"Per Student Budget": school_budget_per_student, 
                                   "Average Math Score": per_school_math, "Average Reading Score": per_school_reading,
                                  "% Passing Math": school_passing_math_perc, "% Passing Reading": school_passing_reading_perc, 
                                   "% Overall Passing": school_passing_both_perc})

# Formating and Mapping school summary DF
per_school_summary["Total School Budget"] = per_school_summary["Total School Budget"].map("${:,.2f}".format)
per_school_summary["Per Student Budget"] = per_school_summary["Per Student Budget"].map("${:,.2f}".format)
per_school_summary["Average Math Score"] = per_school_summary["Average Math Score"].map("{:,.3f}".format)
per_school_summary["Average Reading Score"] = per_school_summary["Average Reading Score"].map("{:,.3f}".format)
per_school_summary["% Passing Math"] = per_school_summary["% Passing Math"].map("{:,.3f}".format)
per_school_summary["% Passing Reading"] = per_school_summary["% Passing Reading"].map("{:,.3f}".format)
per_school_summary["% Overall Passing"] = per_school_summary["% Overall Passing"].map("{:,.3f}".format)


#Displaying school summary DF
per_school_summary.head(15)


School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget Average Math Score  \
Bailey High School               $628.00             77.048   
Cabrera High School              $582.00             83.062   
Figueroa High School             $639.00             76.712   
Ford High School                 $644.00             77.103   
Griffin High School              $625.00             83.351   
Hernandez High School            $652.00             77.290   
Holden High School               $581.00             83.803   
Huang High School                $655.00             76.629   
Johnson High School              $650.00             77.072   
Pena High School                 $609.00             83.840   
Rodriguez High School            $637.00             76.843   
Shelton High School              $600.00             83.359   
Thomas High School               $638.00             83.418   
Wilson High School               $578.00             83.274   
Wright High School               $583.00             83.682   

                      Average Reading Score % Passing Math % Passing Reading  \
Bailey High School                   81.034         66.680            81.933   
Cabrera High School                  83.976         94.133            97.040   
Figueroa High School                 81.158         65.988            80.739   
Ford High School                     80.746         68.310            79.299   
Griffin High School                  83.817         93.392            97.139   
Hernandez High School                80.934         66.753            80.863   
Holden High School                   83.815         92.506            96.253   
Huang High School                    81.183         65.684            81.316   
Johnson High School                  80.966         66.058            81.222   
Pena High School                     84.045         94.595            95.946   
Rodriguez High School                80.745         66.367            80.220   
Shelton High School                  83.726         93.867            95.855   
Thomas High School                   83.849         93.272            97.309   
Wilson High School                   83.989         93.868            96.540   
Wright High School                   83.955         93.333            96.611   

                      % Overall Passing  
Bailey High School               54.642  
Cabrera High School              91.335  
Figueroa High School             53.204  
Ford High School                 54.290  
Griffin High School              90.599  
Hernandez High School            53.528  
Holden High School               89.227  
Huang High School                53.514  
Johnson High School              53.539  
Pena High School                 90.541  
Rodriguez High School            52.988  
Shelton High School              89.892  
Thomas High Schoo

# Highest-Performing Schools (by % Overall Passing)

In [314]:
#Top 5 best performing schools based on overall passing 
top_schools = per_school_summary.sort_values("% Overall Passing", ascending=False)
top_schools.head(5)

School Type  Total Students Total School Budget  \
Cabrera High School     Charter            1858       $1,081,356.00   
Thomas High School      Charter            1635       $1,043,130.00   
Griffin High School     Charter            1468         $917,500.00   
Wilson High School      Charter            2283       $1,319,574.00   
Pena High School        Charter             962         $585,858.00   

                    Per Student Budget Average Math Score  \
Cabrera High School            $582.00             83.062   
Thomas High School             $638.00             83.418   
Griffin High School            $625.00             83.351   
Wilson High School             $578.00             83.274   
Pena High School               $609.00             83.840   

                    Average Reading Score % Passing Math % Passing Reading  \
Cabrera High School                83.976         94.133            97.040   
Thomas High School                 83.849         93.272            97.309   
Griffin High School                83.817         93.392            97.139   
Wilson High School                 83.989         93.868            96.540   
Pena High School                   84.045         94.595            95.946   

                    % Overall Passing  
Cabrera High School            91.335  
Thomas High School             90.948  
Griffin High School            90.599  
Wilson High School             90.583  
Pena High School               90.541

# Bottom Performing Schools (by % Overall Passing)

In [315]:
#Bot 5 worst performing schools based on overall passing
bot_schools = per_school_summary.sort_values("% Overall Passing", ascending=True)
bot_schools.head(5)

School Type  Total Students Total School Budget  \
Rodriguez High School    District            3999       $2,547,363.00   
Figueroa High School     District            2949       $1,884,411.00   
Huang High School        District            2917       $1,910,635.00   
Hernandez High School    District            4635       $3,022,020.00   
Johnson High School      District            4761       $3,094,650.00   

                      Per Student Budget Average Math Score  \
Rodriguez High School            $637.00             76.843   
Figueroa High School             $639.00             76.712   
Huang High School                $655.00             76.629   
Hernandez High School            $652.00             77.290   
Johnson High School              $650.00             77.072   

                      Average Reading Score % Passing Math % Passing Reading  \
Rodriguez High School                80.745         66.367            80.220   
Figueroa High School                 81.158         65.988            80.739   
Huang High School                    81.183         65.684            81.316   
Hernandez High School                80.934         66.753            80.863   
Johnson High School                  80.966         66.058            81.222   

                      % Overall Passing  
Rodriguez High School            52.988  
Figueroa High School             53.204  
Huang High School                53.514  
Hernandez High School            53.528  
Johnson High School              53.539

# Math Scores by Grade

In [316]:
#Seperating data by grade
ninth_graders = school_data_all[(school_data_all["grade"] == "9th")]
tenth_graders = school_data_all[(school_data_all["grade"] == "10th")]
eleventh_graders = school_data_all[(school_data_all["grade"] == "11th")]
twelfth_graders = school_data_all[(school_data_all["grade"] == "12th")]

#Assigning values to each grade
ninth_graders_scores = ninth_graders.groupby(["school_name"])["math_score"].mean()
tenth_graders_scores = tenth_graders.groupby(["school_name"])["math_score"].mean()
eleventh_graders_scores = eleventh_graders.groupby(["school_name"])["math_score"].mean()
twelfth_graders_scores = twelfth_graders.groupby(["school_name"])["math_score"].mean()

#Creating math score DF
math_score_per_grade = pd.DataFrame({"9th": ninth_graders_scores, "10th": tenth_graders_scores, 
                                    "11th": eleventh_graders_scores, "12th":twelfth_graders_scores})

#Minor data wrangling
math_score_per_grade.index.name = None

#Displaying math score DF
math_score_per_grade

9th       10th       11th       12th
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

# Reading Scores by Grade

In [317]:
#Seperating data by grade
ninth_graders = school_data_all[(school_data_all["grade"] == "9th")]
tenth_graders = school_data_all[(school_data_all["grade"] == "10th")]
eleventh_graders = school_data_all[(school_data_all["grade"] == "11th")]
twelfth_graders = school_data_all[(school_data_all["grade"] == "12th")]

#Assigning values to each grade
ninth_graders_rscores = ninth_graders.groupby(["school_name"])["reading_score"].mean()
tenth_graders_rscores = tenth_graders.groupby(["school_name"])["reading_score"].mean()
eleventh_graders_rscores = eleventh_graders.groupby(["school_name"])["reading_score"].mean()
twelfth_graders_rscores = twelfth_graders.groupby(["school_name"])["reading_score"].mean()

#Creating reading score DF
reading_score_per_grade = pd.DataFrame({"9th": ninth_graders_rscores, "10th": tenth_graders_rscores, 
                                    "11th": eleventh_graders_rscores, "12th":twelfth_graders_rscores})

#Minor data wrangling
reading_score_per_grade.index.name = None

#Displaying reading score DF
reading_score_per_grade

9th       10th       11th       12th
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

# Score by School Spending

In [318]:
# Establish the bins & labels
spending_bins = [0.0, 585.0, 630.0, 645.0, 680.0]

labels = ["<$585", "$585-630", "$630-645", "$645-680"]

#Copied the df to a new variable 
school_spending_df = per_school_summary.copy()

#I had some issues with the pd.cut becuase it wasnt taking the value as a float/int because of the "$" so I had to work around
#changed the value to be able to work with the bins.
school_spending_df["Per Student Budget"] = school_spending_df["Per Student Budget"].str.replace("$","", regex=False)
school_spending_df["Per Student Budget"] = school_spending_df["Per Student Budget"].astype(float)


#I had some issues also running the reast of the values since they were obejects so I just changed the typing 
school_spending_df["Average Math Score"] = school_spending_df["Average Math Score"].astype(float)
school_spending_df["Average Reading Score"] = school_spending_df["Average Reading Score"].astype(float)
school_spending_df["% Passing Math"] = school_spending_df["% Passing Math"].astype(float)
school_spending_df["% Passing Reading"] = school_spending_df["% Passing Reading"].astype(float)
school_spending_df["% Overall Passing"] = school_spending_df["% Overall Passing"].astype(float)


#"Per Student Budgets" cataegorized spending based on bins
school_spending_df["Spending Ranges (Per Student)"] = pd.cut(school_spending_df["Per Student Budget"], bins=spending_bins, labels=labels)
#Displayed DF
school_spending_df

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                       Per Student Budget  Average Math Score  \
Bailey High School                  628.0              77.048   
Cabrera High School                 582.0              83.062   
Figueroa High School                639.0              76.712   
Ford High School                    644.0              77.103   
Griffin High School                 625.0              83.351   
Hernandez High School               652.0              77.290   
Holden High School                  581.0              83.803   
Huang High School                   655.0              76.629   
Johnson High School                 650.0              77.072   
Pena High School                    609.0              83.840   
Rodriguez High School               637.0              76.843   
Shelton High School                 600.0              83.359   
Thomas High School                  638.0              83.418   
Wilson High School                  578.0              83.274   
Wright High School                  583.0              83.682   

                       Average Reading Score  % Passing Math  \
Bailey High School                    81.034          66.680   
Cabrera High School                   83.976          94.133   
Figueroa High School                  81.158          65.988   
Ford High School                      80.746          68.310   
Griffin High School                   83.817          93.392   
Hernandez High School                 80.934          66.753   
Holden High School                    83.815          92.506   
Huang High School                     81.183          65.684   
Johnson High School                   80.966          66.058   
Pena High School                      84.045          94.595   
Rodriguez High School                 80.745          66.367   
Shelton High School                   83.726          93.867   
Thomas High School                    83.849          93.272   
Wilson High School                    83.989          93.868   
Wright High School                    83.955          93.333   

                       % Passing Reading  % Overall Passing  \
Bailey High School                81.933             54.642   
Cabrera High School               97.040             91.335   
Figueroa High School              80.739             53.204   
Ford High School                  79.299             54.290   
Griffin High School               97.139             90.599   
Hernandez High School             80.863             53.528   
Holden High School                96.253             89.227   
Huang High School                 81.316             53.514   
Johnson High School               81.222             53.539   
Pena High School                  95.946             90.541   
Rodriguez High School             80.220             52.988   
Shelton High School            

In [319]:
#Calculate averages for the desired columns
spending_math_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean(numeric_only=True)["Average Math Score"]
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean(numeric_only=True)["Average Reading Score"]
spending_passing_math = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean(numeric_only=True)["% Passing Math"]
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean(numeric_only=True)["% Passing Reading"]
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"]).mean(numeric_only=True)["% Overall Passing"]


In [320]:
#DF established
spending_summary =  pd.DataFrame({ "Average Math Score": spending_math_scores,"Average Reading Score": spending_reading_scores,
   "% Passing Math": spending_passing_math,"% Passing Reading": spending_passing_reading,
                                "% Overall Passing": overall_passing_spending,})
#DF displyed
spending_summary.head()

Average Math Score  Average Reading Score  \
Spending Ranges (Per Student)                                              
<$585                                    83.45525              83.933750   
$585-630                                 81.89950              83.155500   
$630-645                                 78.51900              81.624500   
$645-680                                 76.99700              81.027667   

                               % Passing Math  % Passing Reading  \
Spending Ranges (Per Student)                                      
<$585                                93.46000          96.611000   
$585-630                             87.13350          92.718250   
$630-645                             73.48425          84.391750   
$645-680                             66.16500          81.133667   

                               % Overall Passing  
Spending Ranges (Per Student)                     
<$585                                    90.3695  
$585-630                                 81.4185  
$630-645                                 62.8575  
$645-680                                 53.5270

# Scores by School Size

In [321]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

#Pd.cut categorized school size to bins
per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], size_bins, labels=labels)

#Used to convert objects into floats for mean
per_school_summary["Average Math Score"] = per_school_summary["Average Math Score"].astype(float)
per_school_summary["Average Reading Score"] = per_school_summary["Average Reading Score"].astype(float)
per_school_summary["% Passing Math"] = per_school_summary["% Passing Math"].astype(float)
per_school_summary["% Passing Reading"] = per_school_summary["% Passing Reading"].astype(float)
per_school_summary["% Overall Passing"] = per_school_summary["% Overall Passing"].astype(float)

#Displayed DF
per_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976       $3,124,928.00   
Cabrera High School       Charter            1858       $1,081,356.00   
Figueroa High School     District            2949       $1,884,411.00   
Ford High School         District            2739       $1,763,916.00   
Griffin High School       Charter            1468         $917,500.00   
Hernandez High School    District            4635       $3,022,020.00   
Holden High School        Charter             427         $248,087.00   
Huang High School        District            2917       $1,910,635.00   
Johnson High School      District            4761       $3,094,650.00   
Pena High School          Charter             962         $585,858.00   
Rodriguez High School    District            3999       $2,547,363.00   
Shelton High School       Charter            1761       $1,056,600.00   
Thomas High School        Charter            1635       $1,043,130.00   
Wilson High School        Charter            2283       $1,319,574.00   
Wright High School        Charter            1800       $1,049,400.00   

                      Per Student Budget  Average Math Score  \
Bailey High School               $628.00              77.048   
Cabrera High School              $582.00              83.062   
Figueroa High School             $639.00              76.712   
Ford High School                 $644.00              77.103   
Griffin High School              $625.00              83.351   
Hernandez High School            $652.00              77.290   
Holden High School               $581.00              83.803   
Huang High School                $655.00              76.629   
Johnson High School              $650.00              77.072   
Pena High School                 $609.00              83.840   
Rodriguez High School            $637.00              76.843   
Shelton High School              $600.00              83.359   
Thomas High School               $638.00              83.418   
Wilson High School               $578.00              83.274   
Wright High School               $583.00              83.682   

                       Average Reading Score  % Passing Math  \
Bailey High School                    81.034          66.680   
Cabrera High School                   83.976          94.133   
Figueroa High School                  81.158          65.988   
Ford High School                      80.746          68.310   
Griffin High School                   83.817          93.392   
Hernandez High School                 80.934          66.753   
Holden High School                    83.815          92.506   
Huang High School                     81.183          65.684   
Johnson High School                   80.966          66.058   
Pena High School                      84.045          94.595   
Rodriguez High School                 80.745          66.367   
Shelton High School                   83.726          93.867   
Thomas High School                    83.849          93.272   
Wilson High School                    83.989          93.868   
Wright High School                    83.955          93.333   

                       % Passing Reading  % Overall Passing  \
Bailey High School                81.933             54.642   
Cabrera High School               97.040             91.335   
Figueroa High School              80.739             53.204   
Ford High School                  79.299             54.290   
Griffin High School               97.139             90.599   
Hernandez High School             80.863             53.528   
Holden High School                96.253             89.227   
Huang High School                 81.316             53.514   
Johnson High School               81.222             53.539   
Pena High School                  95.946             90.541   
Rodriguez High School             80.220             52.988   
Shelton High School               95.855       

In [322]:
# Calculate averages for the desired columns. 
size_math_scores = per_school_summary.groupby(["School Size"]).mean(numeric_only=True)["Average Math Score"]
size_reading_scores = per_school_summary.groupby(["School Size"]).mean(numeric_only=True)["Average Reading Score"]
size_passing_math = per_school_summary.groupby(["School Size"]).mean(numeric_only=True)["% Passing Math"]
size_passing_reading = per_school_summary.groupby(["School Size"]).mean(numeric_only=True)["% Passing Reading"]
size_overall_passing = per_school_summary.groupby(["School Size"]).mean(numeric_only=True)["% Overall Passing"]

#DF was created
size_summary =  pd.DataFrame({ "Average Math Score": size_math_scores,"Average Reading Score": size_reading_scores,
   "% Passing Math": size_passing_math,"% Passing Reading": size_passing_reading,
                                "% Overall Passing": size_overall_passing,})
#DF was displayed
size_summary

Average Math Score  Average Reading Score  % Passing Math  \
School Size                                                                     
Small (<1000)                83.821500              83.930000         93.5505   
Medium (1000-2000)           83.374400              83.864600         93.5994   
Large (2000-5000)            77.746375              81.344375         69.9635   

                    % Passing Reading  % Overall Passing  
School Size                                               
Small (<1000)                 96.0995            89.8840  
Medium (1000-2000)            96.7908            90.6214  
Large (2000-5000)             82.7665            58.2860

# Scores by School Type

In [323]:
# Calculate averages for the desired columns.
type_math_scores = per_school_summary.groupby(["School Type"]).mean(numeric_only=True)["Average Math Score"]
type_reading_scores = per_school_summary.groupby(["School Type"]).mean(numeric_only=True)["Average Reading Score"]
type_passing_math = per_school_summary.groupby(["School Type"]).mean(numeric_only=True)["% Passing Math"]
type_passing_reading = per_school_summary.groupby(["School Type"]).mean(numeric_only=True)["% Passing Reading"]
type_overall_passing = per_school_summary.groupby(["School Type"]).mean(numeric_only=True)["% Overall Passing"]

#DF was created
type_summary = pd.DataFrame({ "Average Math Score": type_math_scores,"Average Reading Score": type_reading_scores,
   "% Passing Math": type_passing_math,"% Passing Reading": type_passing_reading,
                                "% Overall Passing": type_overall_passing,})

#DF was displayed
type_summary


Average Math Score  Average Reading Score  % Passing Math  \
School Type                                                              
Charter               83.473625              83.896500       93.620750   
District              76.956714              80.966571       66.548571   

             % Passing Reading  % Overall Passing  
School Type                                        
Charter              96.586625          90.432250  
District             80.798857          53.672143

# Written Report

- Summary:

    _The school district as a total invested $24,649,428.00 million dollars into its schools this year. The results produced as a school district were a math passing average of 74.981% and a reading passing average of 85.805% with an overall passing average of 65.172%. The schools individually that performed best out of the 15 schools in our district were Cabrera High School, Thomas High School, Griffin High School, Wilson High School, and Pena High School. They all have a student passing rate of over at least 93% in reading and math and an overall passing rate of at least 90%. The worst performing schools in the district were Rodriguez High School, Figueroa High School, Huang High School, Hernandez High School, and Johnson High School. They all have a student passing math rate that is barely above 66%, reading passing rate at just below 81% and an overall passing rate of just barely 53%. The most amount of money that is being allocated per student is mainly on the lowest performing schools, but they are also the schools with the most students. The discrepancy between Charter and District schools is quite massive in both passing math percentage and overall passing percentage._

- Conclusions:

     _Charter schools have surpassed District schools in every academic category that was measured. The size of the students could be a reason why they are not achieving better results. Schools with medium to small school sizes all produced better results; with that said, a charter large size school was able to produce the same results that small to medium schools was able to produce. This could be an issue with how the district schools are being operated or the student to staff ratio is a factor for the results, I would look into these areas to see if we could improve the district school academic performance._